immutablecollections/_immutabledict.py

```diff
@@ -228,8 +233,16 @@ def __contains__(self, x: object) -> bool:
         return self._dict.__contains__(x)
 
     def __hash__(self) -> int:
-        return hash(self._dict)
+        # This hashing implementation is borrowed from frozendict:
+        # https://github.com/slezica/python-frozendict/blob/c5d16bafcca7b72ff3e8f40d3a9081e4c9233f1b/frozendict/__init__.py#L46
+        if self._hash is None:
```

Comment:
Can you benchmark whether or not it is faster to construct a `set` of key-value pairs and use that set's hash as the hash value?

Reply:
``` python
from immutablecollections import immutabledict
from copy import copy
import random


empty_tuple = ()
empty_list = []
empty_dict = {}

small_tuple = (3, 1, 2)
small_list = [3, 1, 2]
small_dict = {"a": 3, "b": 1, "c": 2}

rand = random.Random(0)

big_dim = int(1e3)

big_list = list(range(big_dim))
rand.shuffle(big_list)
big_tuple = tuple(big_list)
rand.shuffle(big_list)
big_dict = dict(zip(big_list, big_tuple))

empty_immutabledict = immutabledict()
small_immutabledict = immutabledict(small_dict)
big_immutabledict = immutabledict(big_dict)

# ensure immutable dictionaries are all different to force hash computation
empty_dicts = [copy(empty_immutabledict) for _ in range(big_dim)]
small_dicts = [copy(small_immutabledict) for _ in range(big_dim)]
big_dicts = [copy(big_immutabledict) for _ in range(big_dim)]
assert len(set(id(d) for d in empty_dicts)) == 1
assert len(set(id(d) for d in small_dicts)) == len(small_dicts)
assert len(set(id(d) for d in big_dicts)) == len(big_dicts)

print("empty")
%timeit [hash(d) for d in empty_dicts]
print("small")
%timeit [hash(d) for d in small_dicts]
print("big")
%timeit [hash(d) for d in big_dicts]
```
Implementation with hashing based on repeated xor:
```
empty
149 µs ± 2.52 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
small
168 µs ± 1.37 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
big
174 µs ± 1.51 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
Implementation with hashing a tuple of KV pairs:

```
empty
166 µs ± 1.52 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
small
176 µs ± 1.16 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
big
179 µs ± 1.4 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
Implementation with hashing an immutable set of KV pairs:

```
empty
158 µs ± 854 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
small
163 µs ± 337 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
big
165 µs ± 1.1 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Reply:
Implementation with hashing a `frozenset` of KV pairs:

```
empty
163 µs ± 1.66 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
small
172 µs ± 4.99 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
big
172 µs ± 1.91 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```